### ISSUE TITLE
intercepting the file before download

### ISSUE THREAD

**Issue by ofZach** [tapioca24/p5.capture#16]
great library!  

forgive me if this is obvious but in the case that I wanted to let people review the video before downloading, would there be an elegant way to interrupt the process so that the file isn't downloaded directly?   at the moment, the file just downloads but in some cases, it may be helpful to preview the recorded video as the user may want to discard it (this is for an interactive work).  My understanding / memory is a little hazy, but I guess it's a blob before it gets downloaded.  

thanks much in advance!

**Comment by guilhermesilveira** [comment]
Yes, you are right, you can get a hold of the blob before downloading. The blob is generated inside each recorder:
https://github.com/tapioca24/p5.capture/blob/0cf2f2e4359c45e99944a30412c1518928c78a7f/src/recorders/webm-recorder.ts#L66

The event listener which receives the blob and downloads it:
https://github.com/tapioca24/p5.capture/blob/8b63af83179473939b89b13ea9882038a1b68cd1/src/p5.capture.ts#L285

Depending on your config you might want a segmented event handler too.

The easiest path is to add an option for a customized onFinish callback, which gets called with (blob, filename, defaultHandler) instead of the download. To keep it uniform, perhaps allow for all event handlers to be configured through the options (onFinished, onProgress, onAdded, onSegmented, onError, onStop, onStart).

It would look like:

```
P5Capture.setDefaultOptions({
  onFinished: (blob, filename, defaultHandler) => {
    // do your preview and then invoke defaultHandler later if you want to
  }
});
```

**Comment by tapioca24** [comment]
Hi there!
Thank you @ofZach for your request.
Thank you @guilhermesilveira for pointing me in the right direction.

The callback option to hook into events is flexible and powerful and I agree with this direction.
I plan to reorganize the interface a bit and add the following option.

```ts
P5Capture.setDefaultOptions({
  beforeDownload: (blob, context, next) => {
    // do something...
    next()
  }
})
```

type is as follows.

```ts
type P5CaptureOptions = {
  // ...
  beforeDownload?: (
    blob: Blob,
    context: {
      filename: string;
      format: P5CaptureFormat;
      isSegmented: boolean;
    },
    next: () => void,
  ) => Promise<void> | void;
};
```

The `context` object provides a source of information for additional processing.
Calling the `next` callback will start the file download, but can be omitted if not needed.

If you have any concerns, we would love to hear your feedback.

**Comment by tapioca24** [comment]
Released in v1.4.0. Thank you for all your contributions!
